File: src/actions/command.cpp
#include "actions/command.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Actions
{
Command::Command(const PbActions::Command& protobuf) {
    input = protobuf.input();
    output = protobuf.output();
}

Command::operator PbActions::Command() const {
    PbActions::Command ret;
    ret.set_input(input);
    ret.set_output(output);
    return ret;
}

std::string Command::serializeAsJsonString() const {
    PbActions::Command protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string Command::serializeAsProtobufString() const {
    PbActions::Command protobuf(*this);
    return protobuf.SerializeAsString();
}

bool Command::deserializeFromJsonString(const std::string& str) {
    PbActions::Command protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool Command::deserializeFromProtobufString(const std::string& str) {
    PbActions::Command protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}

File: py/actions/command.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from actions import command_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class Command:
    input: str = ""
    output: str = ""
    
    _proto_message: command_pb2.Command = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = command_pb2.Command()

    def _populate_proto(self):
        self._proto_message.input = self.input
        self._proto_message.output = self.output

    @classmethod
    def _from_proto(cls, proto_message) -> "Command":
        return cls(
            input = proto_message.input,
            output = proto_message.output,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "Command":
        message = command_pb2.Command()
        message.ParseFromString(data)
        return cls(
            input = message.input,
            output = message.output,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "Command":
        message = command_pb2.Command()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())
